waveorder/waveorder_reconstructor.py

```diff
@@ -977,23 +988,40 @@ def Stokes_recon(self, I_meas):
         Parameters
         ----------
             I_meas        : numpy.ndarray
-                            polarization-sensitive intensity images with the size of (N_channel, ...)
+                            polarization-sensitive intensity images with the size of (N_channel, N, M) or
+                            (N_channel, N, M, N_defocus)
                           
         Returns
         -------
             S_image_recon : numpy.ndarray
-                            reconstructed Stokes parameters with the size of (N_Stokes, ...)
+                            reconstructed Stokes parameters with the size of (N_Stokes, N, M) or
+                            (N_channel, N, M, N_defocus)
                        
                               
         '''
-        
-        img_shape = I_meas.shape
-        
-        A_pinv = np.linalg.pinv(self.A_matrix)
-        S_image_recon = np.reshape(np.dot(A_pinv, I_meas.reshape((self.N_channel, -1))), (self.N_Stokes,)+img_shape[1:])
 
-            
-        return S_image_recon
+        n_dims = I_meas.ndim
+        if n_dims not in (3, 4):
+            raise ValueError('Unsupported image data size. Image data must be of size (N_channel, N, M) or '
+                             '(N_channel, N, M, N_defocus)')
+
+        if n_dims == 3:
+            I_meas = I_meas[..., np.newaxis]
+        img_data = np.moveaxis(I_meas, 0, -2)
+
+        if self.use_gpu:
+            if self._A_matrix_inv_gpu_array is None:
+                self._A_matrix_inv_gpu_array = cp.array(self.A_matrix_inv)
+            img_gpu_array = cp.array(img_data)
+            S_image_recon = cp.asnumpy(cp.matmul(self._A_matrix_inv_gpu_array, img_gpu_array))
+        else:
+            S_image_recon = np.matmul(self.A_matrix_inv, img_data)
+
+        S_image_recon = np.moveaxis(S_image_recon, -2, 0)
+        if n_dims == 3:
+            return S_image_recon[..., 0]
+        else:
+            return S_image_recon
```

Comment:
A potential solution that can handle both instrument matrix and instrument tensor cases is to vectorize only XZ dimensions (i=Z*X*Y) in intensity, stokes, and instrument tensor arrays. In this case, intensities will be of (i,4) shape for both 2D images and 3D stacks. Correspondingly, Stokes array will be (i,3) shape. The inverse instrument tensor can have the shape (1,4,3) OR (i,4,3). 

Based on the broadcasting behavior of numpy.matmul (noted below), it seems like np.matmul or cp.matmul will multiply arrays of shape (i,4) and (i,4,3) to return an array of shape (i,3). 

(https://numpy.org/doc/stable/reference/generated/numpy.matmul.html, "If either argument is N-D, N > 2, it is treated as a stack of matrices residing in the last two indexes and broadcast accordingly.")

Reply:
This could work
```
import numpy as np
nVox=100
inst_mat=np.array([[0.25,0.25,0.25,0.25],[1,0,-1,0],[0,1,0,-1]])
inst_mat=inst_mat.reshape(1,3,4)
print("inst_mat:")
print(inst_mat)

inst_tensor=np.repeat(inst_mat,nVox,axis=0)
intensity=np.random.rand(nVox,4,1)
print("instrument tensor:"+str(inst_tensor.shape))
print("intensity shape:"+ str(intensity.shape))
stokes = np.matmul(inst_tensor, intensity)
print("stokes shape:" + str(stokes.shape))
```

produces

```
inst_mat:
[[[ 0.25  0.25  0.25  0.25]
  [ 1.    0.   -1.    0.  ]
  [ 0.    1.    0.   -1.  ]]]
instrument tensor:(100, 3, 4)
intensity shape:(100, 4, 1)
stokes shape:(100, 3, 1)
```

Reply:
Thanks Shalin, the issue with the current implementation is that the function needs to accept data of formats `N_channel, N, M`, `N_channel, N_pattern, N, M`, `N_channel, N, M, N_defocus`, `N_channel, N_pattern, N, M, N_defocus`, and its ambiguous where the image dimensions are. The previous implementation used to reshape I_mean into `(N_channel, -1)` and then multiply by the instrument matrix, this strategy does not work in the case of the instrument tensor